File: rag-service/app/core/retriever.py
"""Recuperação dos excertos mais relevantes para uma pergunta."""

from dataclasses import dataclass

import httpx

from app.core.db import get_supabase, rows_of, run_db
from app.core.embeddings import generate_embedding

DEFAULT_K = 5


@dataclass(frozen=True, slots=True)
class RetrievedChunk:
    """Chunk devolvido pela pesquisa vectorial, já com o score de similaridade."""

    id: str
    file_path: str
    content: str
    start_offset: int
    end_offset: int
    chunk_index: int
    similarity: float


async def retrieve_chunks(
    client: httpx.AsyncClient,
    question: str,
    repo_id: str,
    k: int = DEFAULT_K,
) -> list[RetrievedChunk]:
    """Devolve os top-k chunks do repo, por similaridade decrescente.

    Encapsula a geração do embedding da pergunta e a chamada à função SQL
    `match_chunks` via RPC.
    """
    query_embedding = await generate_embedding(client, question)

    rows = await run_db(
        lambda: rows_of(
            get_supabase()
            .rpc(
                "match_chunks",
                {
                    "query_embedding": query_embedding,
                    "match_repo_id": repo_id,
                    "match_count": k,
                },
            )
            .execute()
        ),
        "correr a função match_chunks",
    )

    return [
        RetrievedChunk(
            id=str(row["id"]),
            file_path=row["file_path"],
            content=row["content"],
            start_offset=row["start_offset"],
            end_offset=row["end_offset"],
            chunk_index=row["chunk_index"],
            similarity=float(row["similarity"]),
        )
        for row in rows
    ]

File: apps/web/lib/retriever.ts
import { createClient } from "@supabase/supabase-js";
import { generateEmbedding } from "./gemini";

const DEFAULT_K = 5;

export type RetrievedChunk = {
  id: string;
  file_path: string;
  content: string;
  start_offset: number;
  end_offset: number;
  chunk_index: number;
  similarity: number;
};

function getSupabaseClient() {
  const url = process.env.SUPABASE_URL;
  const key = process.env.SUPABASE_SERVICE_ROLE_KEY;
  if (!url || !key) {
    throw new Error(
      "SUPABASE_URL ou SUPABASE_SERVICE_ROLE_KEY não estão definidos no .env.local"
    );
  }
  return createClient(url, key);
}

// Recebe uma pergunta em linguagem natural e devolve os top-k chunks do repo
// ordenados por similaridade decrescente. Encapsula geração do embedding da
// query + chamada à função SQL match_chunks via RPC.
export async function retrieveChunks(
  question: string,
  repoId: string,
  k: number = DEFAULT_K
): Promise<RetrievedChunk[]> {
  const supabase = getSupabaseClient();

  const queryEmbedding = await generateEmbedding(question);

  const { data, error } = await supabase.rpc("match_chunks", {
    query_embedding: queryEmbedding,
    match_repo_id: repoId,
    match_count: k,
  });

  if (error) {
    throw new Error(`Falha na função match_chunks: ${error.message}`);
  }

  return (data ?? []) as RetrievedChunk[];
}
